### ISSUE TITLE
Bad frames generated when using composition of affine filtered video and webvfx

### ISSUE THREAD

**Issue by geoffroymontel** [mltframework/mlt#109]
Hello

After my last issue https://github.com/mltframework/mlt/issues/108, I removed QT4, installed QT5, and installed MELT V6.0.0

```bash
$ melt -version
melt 6.0.0
Copyright (C) 2002-2016 Meltytech, LLC
<http://www.mltframework.org/>
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
$ ffmpeg
ffmpeg version N-79294-g972df59 Copyright (c) 2000-2016 the FFmpeg developers
  built with gcc 4.8 (Ubuntu 4.8.4-2ubuntu1~14.04.1)
  configuration: --prefix=/opt/melt --enable-gpl --enable-version3 --enable-shared --enable-debug --enable-pthreads --enable-runtime-cpudetect --disable-doc --enable-libtheora --enable-libvorbis --enable-libmp3lame --enable-libx264 --enable-libx265 --enable-libvpx
  libavutil      55. 20.100 / 55. 20.100
  libavcodec     57. 34.100 / 57. 34.100
  libavformat    57. 32.100 / 57. 32.100
  libavdevice    57.  0.101 / 57.  0.101
  libavfilter     6. 41.100 /  6. 41.100
  libswscale      4.  1.100 /  4.  1.100
  libswresample   2.  0.101 /  2.  0.101
  libpostproc    54.  0.100 / 54.  0.100
```

But I have a regression (compared to Melt 0.9.6 running with QT4) on a specific example : bad frames are generated when using a combination of webvfx and affine filtered videos.

## Example where bad frames are generated

MELT file

```xml
<mlt LC_NUMERIC="en_US.UTF-8">
  <producer id="1">
    <property name="resource">../../melt-tests/penguin.mp4</property>
  </producer>
  <producer id="63">
    <property name="mlt_service">webvfx</property>
    <property name="resource">plain:boc-bug.html</property>
    <property name="transparent">1</property>
  </producer>
  <tractor>
    <multitrack>
      <tractor>
        <multitrack>
          <tractor>
            <playlist>
              <entry producer="1" in="0" out="200">
              </entry>
            </playlist>
          </tractor>
        </multitrack>
        <filter>
          <property name="mlt_service">affine</property>
          <property name="transition.cycle">25</property>
          <property name="transition.geometry">0=0/0:110%x110%;22=0/0:110%x110%;23=0/-2:110%x110%;24=0/0:110%x110%</property>
          <property name="transition.distort">0</property>
        </filter>
      </tractor>
      <tractor>
        <multitrack>
          <tractor>
            <playlist>
              <entry producer="63" in="0" out="124">
              </entry>
            </playlist>
          </tractor>
        </multitrack>
      </tractor>
    </multitrack>
    <transition in="0" out="99">
      <property name="mlt_service">composite</property>
      <property name="a_track">0</property>
      <property name="b_track">1</property>
    </transition>
    <transition in="100" out="124">
      <property name="mlt_service">luma</property>
      <property name="a_track">0</property>
      <property name="b_track">1</property>
      <property name="reverse">1</property>
    </transition>
  </tractor>
</mlt>
```

HTML file 

```html
<html>
<head>
</head>
<body>
  <h1 style="color: #FFF">TITLE TITLE TITLE TITLE TITLE TITLE TITLE TITLE TITLE TITLE TITLE TITLE TITLE</h1>
</body>
</html>
```

Profile my_nhd

```
description=my nHD
frame_rate_num=25
frame_rate_den=1
width=640
height=360
progressive=0
sample_aspect_num=1
sample_aspect_den=1
display_aspect_num=16
display_aspect_den=9
colorspace=709
```

Command 

```bash
$ xvfb-run -a melt -profile my_nhd boc-bug.xml -progress -consumer avformat:boc-bug.mp4 strict=experimental
```

Output
![capture d ecran 2016-04-11 a 17 25 12](https://cloud.githubusercontent.com/assets/352910/14432468/a5437de8-000a-11e6-8c6d-ad48c1e9977f.png)

## Removing affine filter

Same HTML

Affine filter is commented

```xml
<mlt LC_NUMERIC="en_US.UTF-8">
  <producer id="1">
    <property name="resource">../../melt-tests/penguin.mp4</property>
  </producer>
  <producer id="63">
    <property name="mlt_service">webvfx</property>
    <property name="resource">plain:boc-bug.html</property>
    <property name="transparent">1</property>
  </producer>
  <tractor>
    <multitrack>
      <tractor>
        <multitrack>
          <tractor>
            <playlist>
              <entry producer="1" in="0" out="200">
              </entry>
            </playlist>
          </tractor>
        </multitrack>
        <!-- <filter>
          <property name="mlt_service">affine</property>
          <property name="transition.cycle">25</property>
          <property name="transition.geometry">0=0/0:110%x110%;22=0/0:110%x110%;23=0/-2:110%x110%;24=0/0:110%x110%</property>
          <property name="transition.distort">0</property>
        </filter> -->
      </tractor>
      <tractor>
        <multitrack>
          <tractor>
            <playlist>
              <entry producer="63" in="0" out="124">
              </entry>
            </playlist>
          </tractor>
        </multitrack>
      </tractor>
    </multitrack>
    <transition in="0" out="99">
      <property name="mlt_service">composite</property>
      <property name="a_track">0</property>
      <property name="b_track">1</property>
    </transition>
    <transition in="100" out="124">
      <property name="mlt_service">luma</property>
      <property name="a_track">0</property>
      <property name="b_track">1</property>
      <property name="reverse">1</property>
    </transition>
  </tractor>
</mlt>
```

Video is perfect in this case
![capture d ecran 2016-04-11 a 17 27 57](https://cloud.githubusercontent.com/assets/352910/14432560/0187661e-000b-11e6-97e6-b60c124ef991.png)

Many thanks again in advance for your help

**Comment by bmatherly** [comment]
I don't know the exact cause of your problem, but it might help to debug if you simplify your XML.

1) Remove all instances of <multitrack>
2) You should only need one <tractor> - and I don't think you need to nest <tractor>s
3) Try to boil your XML down to be as simple as it can be.

It might help you if you use Shotcut to create a simple example of what you want to want and then modify/extrapolate from that.

**Comment by geoffroymontel** [comment]
Dear Brian

Thanks for your quick reply.

I tried to boil my XML down and it's getting worse ;)
I think I still need a multitrack because I'm compositing two tracks.

This XML makes either melt crash or generate bad frames (I can launch it several times and it does not crash at the same percentage and sometimes does not crash)

```xml
<mlt LC_NUMERIC="en_US.UTF-8">
  <producer id="1">
    <property name="resource">video.mp4</property>
  </producer>
  <producer id="63">
    <property name="mlt_service">webvfx</property>
    <property name="resource">plain:boc-bug.html</property>
    <property name="transparent">1</property>
  </producer>
  <tractor>
    <multitrack>
      <playlist>
        <entry producer="1" in="0" out="200">
        </entry>
      </playlist>
      <playlist>
        <entry producer="63" in="0" out="124">
        </entry>
      </playlist>
    </multitrack>
    <transition in="0" out="99">
      <property name="mlt_service">composite</property>
      <property name="a_track">0</property>
      <property name="b_track">1</property>
    </transition>
    <transition in="100" out="124">
      <property name="mlt_service">luma</property>
      <property name="a_track">0</property>
      <property name="b_track">1</property>
      <property name="reverse">1</property>
    </transition>
  </tractor>
</mlt>
```

I'm pretty sure this has something to do with using QT5.
I will investigate more and keep you updated of my results.

All the best

Geoffroy

**Comment by bmatherly** [comment]
Another suggestion: Figure out how to create the effect you are looking for using the melt command line application. When it looks right, use the xml consumer to export as XML. Then, you can inspect the XML.

If you can provide a melt command that reproduces the problem (without using an XML producer), then I might be able to look at it.

**Comment by geoffroymontel** [comment]
OK ! I am currently trying different versions of QT, FFMPEG and MELT to try and find where it breaks, but using the melt command line is a good idea. Thanks Brian !

**Comment by geoffroymontel** [comment]
Dear Brian

I followed your advice and tried the command line.
I'm running a fresh install of Ubuntu 14.04TLS
I've just compiled mlt from https://github.com/mltframework/mlt master branch

Here are the packages I installed before compiling : 
```
"git", "automake", "autoconf", "libtool", "intltool", "g++", "yasm", "swig", "libmp3lame-dev",
  "libgavl-dev", "libsamplerate-dev", "libxml2-dev", "ladspa-sdk", "libjack-dev", "libsox-dev", "libsdl-dev",
  "libgtk2.0-dev", "liboil-dev", "libsoup2.4-dev", "libexif-dev", "libtheora-dev", "libvdpau-dev",
  "libvorbis-dev", "python-dev", "cmake", "libeigen3-dev", "libglew-dev", "xutils-dev", "libegl1-mesa-dev", "libfftw3-dev",
  'qt5-default','libqt5webkit5-dev','qtdeclarative5-dev','qtquick1-5-dev','qtscript5-dev','libqt5svg5-dev'
```

Here is the `melt.conf` file used when compiling melt

```
INSTALL_DIR="/opt/melt"
AUTO_APPEND_DATE=0
SOURCE_DIR="/opt/melt/src"
MLT_HEAD=1
ENABLE_WEBVFX=1
FFMPEG_HEAD=1
```

So that 

```bash
$ melt -version
melt 6.1.0
Copyright (C) 2002-2016 Meltytech, LLC
<http://www.mltframework.org/>
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
$ ffmpeg
ffmpeg version N-79468-ga676c29 Copyright (c) 2000-2016 the FFmpeg developers
  built with gcc 4.8 (Ubuntu 4.8.4-2ubuntu1~14.04.1)
  configuration: --prefix=/opt/melt --enable-gpl --enable-version3 --enable-shared --enable-debug --enable-pthreads --enable-runtime-cpudetect --disable-doc --enable-libtheora --enable-libvorbis --enable-libmp3lame --enable-libx264 --enable-libx265 --enable-libvpx
  libavutil      55. 22.100 / 55. 22.100
  libavcodec     57. 34.102 / 57. 34.102
  libavformat    57. 34.101 / 57. 34.101
  libavdevice    57.  0.101 / 57.  0.101
  libavfilter     6. 42.100 /  6. 42.100
  libswscale      4.  1.100 /  4.  1.100
  libswresample   2.  0.101 /  2.  0.101
  libpostproc    54.  0.100 / 54.  0.100
Hyper fast Audio and Video encoder
usage: ffmpeg [options] [[infile options] -i infile]... {[outfile options] outfile}...
```

This command line works and the video is correct

```bash
$ xvfb-run -a melt video.mp4 in=0 out=200 \
  -track \
  color:#80224466 in=0 out=124 \
  -transition composite in=0 out=99 a_track=0 b_track=1 \
  -progress -profile my_nhd -consumer avformat:boc-bug.mp4 strict=experimental
[libx264 @ 0x7fd6a80c7a80] interlace + weightp is not implemented
[mp4 @ 0x7fd6a80c6da0] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
[mp4 @ 0x7fd6a80c6da0] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
Current Frame:        200, percentage:         99
```

This command line works and the video is correct

```bash
$ xvfb-run -a melt video.mp4 in=0 out=200 \
  -track \
  webvfx:plain:boc-bug.html in=0 out=124 transparent=1 \
  -progress -profile my_nhd -consumer avformat:boc-bug.mp4 strict=experimental
WARNING: QApplication was not created in the main() thread.
[libx264 @ 0x7fd2780c7a80] interlace + weightp is not implemented
[mp4 @ 0x7fd2780c6da0] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
[mp4 @ 0x7fd2780c6da0] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
Current Frame:        200, percentage:         99
```

This command line crashes melt

```bash
$ xvfb-run -a melt video.mp4 in=0 out=200 \
>   -track \
>   webvfx:plain:boc-bug.html in=0 out=124 transparent=1 \
>   -transition composite in=0 out=99 a_track=0 b_track=1 \
>   -progress -profile my_nhd -consumer avformat:boc-bug.mp4 strict=experimental
WARNING: QApplication was not created in the main() thread.
[libx264 @ 0x7f3d900c7a80] interlace + weightp is not implemented
[mp4 @ 0x7f3d900c6da0] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
[mp4 @ 0x7f3d900c6da0] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
Segmentation fault (core dumped)ntage:          1
```

qmelt also crashes

```bash
$ xvfb-run -a qmelt video.mp4 in=0 out=200 \
>   -track \
>   webvfx:plain:boc-bug.html in=0 out=124 transparent=1 \
>   -transition composite in=0 out=99 a_track=0 b_track=1 \
>   -progress -profile my_nhd -consumer avformat:boc-bug.mp4 strict=experimental
[libx264 @ 0x7f91f00c7a80] interlace + weightp is not implemented
[mp4 @ 0x7f91f00c6da0] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
[mp4 @ 0x7f91f00c6da0] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
Segmentation fault (core dumped)ntage:          1
```

Many thanks again for your time and help, all the best

Geoffroy

**Comment by ddennedy** [comment]
I reproduced this using the above command line and qmelt from Shotcut 16.04. It crashes here:
`#0  convert_image (frame=0x7fff9c804260, image=0x7fff377fd928, format=0x7fff377fd7ac, output_format=mlt_image_yuv422) at filter_avcolour_space.c:194
`
I also reproduced the corrupt video when using the affine or frei0r.cairoblend transitions. I do not reproduce it when using `color:#9900ff00` instead of the webvfx producer (color is also rgba). This points to something wrong with the way the webvfx producer is providing images.
